diff --git a/model.py b/model.py
index a89defd..11b1285 100644
--- a/model.py
+++ b/model.py
@@ -13,7 +13,9 @@ from funasr.train_utils.device_funcs import force_gatherable
 from funasr.losses.label_smoothing_loss import LabelSmoothingLoss
 from funasr.metrics.compute_acc import compute_accuracy, th_accuracy
 from funasr.utils.load_utils import load_audio_text_image_video, extract_fbank
-from utils.ctc_alignment import ctc_forced_align
+import sys
+sys.path.append('/home/daoyuan_mm/data-juicer/thirdparty/humanvbench_models')
+from SenseVoice.utils.ctc_alignment import ctc_forced_align
 
 class SinusoidalPositionEncoder(torch.nn.Module):
     """ """
